### PR TITLE
Fix data breakpoint issues

### LINKED ISSUE #46644: Managed Data Breakpoints no longer work against .NET 5+
Due to deadlocking issues with Before/After GC events, the runtime was changed in .NET 5+ to allow a debugger to use Pinned Handles in order to support managed data bp's as opposed to GC Events.

After this change was made in the runtime, VS Debugger found issues with the approach: objects with pinned handles still cause writes to "data bp tracked" memory regions during GC. These writes cause the harware data bp exceptions to be fired during GC, which results in the CLR process crashing for (currently) unknown reasons.

We need a fix from CLR to support Managed Data Breakpoints going forward, ideally one that can be submitted for 5.0 servicing.

### PR DESCRIPTION
**Pull request by sdmaclea**
Currently being tested

Fixes #46644

/cc @chuckries

### CODE REVIEW

File: src/coreclr/debug/ee/controller.cpp [hunk 1]
```diff
@@ -9054,6 +9054,23 @@ bool DebuggerDataBreakpoint::TriggerSingleStep(Thread *thread, const BYTE *ip)
         this->DisableSingleStep();
         return true;
     }
+    else if (g_pDebugger->m_isSuspendedForGarbageCollection)
+    {
+        // The debugger is not interested in Data Breakpoints during garbage collection
+        // We can safely ignore them since the Data Breakpoints are now on pinned objects
+        LOG((LF_CORDB, LL_INFO10000, "D:DDBP: Ignoring data breakpoint while in SuspendEE() for GC \n"));
+
+        // We enabled single step in TriggerDataBreakpoint so that the ScanForTriggers
```
Comment (sdmaclea) [inline src/coreclr/debug/ee/controller.cpp:9063]: Refers to https://github.com/dotnet/runtime/blob/09a1e77f440b478e212195f0d6a4c8fa2564e825/src/coreclr/debug/ee/controller.cpp#L9020-L9022 above

File: src/coreclr/debug/ee/controller.cpp [hunk 2]
```diff
@@ -9054,6 +9054,23 @@ bool DebuggerDataBreakpoint::TriggerSingleStep(Thread *thread, const BYTE *ip)
...
+        // would return DPOSS_USED_WITH_NO_EVENT effectively treating the data breakpoint
```
Comment (sdmaclea) [inline src/coreclr/debug/ee/controller.cpp:9064]: https://github.com/dotnet/runtime/blob/09a1e77f440b478e212195f0d6a4c8fa2564e825/src/coreclr/debug/ee/controller.cpp#L2735-L2741

**Comment by sdmaclea** [comment]
The alternative design would to have `TriggerDataBreakpoint` return `true` when we are in `SuspendEE` and hit a DBP.  Then modify this code to not create the `DebuggerDataBreakpoint` but set `DPOSS_USED_WITH_NO_EVENT` instead.

https://github.com/dotnet/runtime/blob/09a1e77f440b478e212195f0d6a4c8fa2564e825/src/coreclr/debug/ee/controller.cpp#L2708-L2716

I felt this PR might be easier to get through servicing for .NET 5.  In my mind it is slightly simpler.

**Comment by noahfalk** [comment]
For patching .NET 5 this looks like a good fix that minimizes the change in behavior from what we have right now. I'm hoping in the longer term fix for .NET 6 we could be a little more aggresive and avoid ever creating/queueing the DebuggerDataBreakpoint controller during the GC. It looks like the current fix is counting on us never going down the JIT write barrier path while the GC is running. Certainly I'd be very surprised if we called the JIT helper from within the GC but it feels more satisfying if long term we didn't have to rely on that kind of assumption.

**Review by noahfalk — APPROVED** [review]
LGTM!

File: src/coreclr/debug/ee/controller.cpp [hunk 3]
```diff
@@ -9054,6 +9054,23 @@ bool DebuggerDataBreakpoint::TriggerSingleStep(Thread *thread, const BYTE *ip)
         this->DisableSingleStep();
         return true;
     }
+    else if (g_pDebugger->m_isSuspendedForGarbageCollection)
+    {
+        // The debugger is not interested in Data Breakpoints during garbage collection
+        // We can safely ignore them since the Data Breakpoints are now on pinned objects
+        LOG((LF_CORDB, LL_INFO10000, "D:DDBP: Ignoring data breakpoint while in SuspendEE() for GC \n"));
```
Comment (cshung) [inline src/coreclr/debug/ee/controller.cpp:9061]: Technically, we need to ignore it not just during `SuspendEE()`, but throughout the whole suspension.

```suggestion
        LOG((LF_CORDB, LL_INFO10000, "D:DDBP: Ignoring data breakpoint while in the runtime is suspended for GC \n"));
```

**Review by cshung — APPROVED** [review]
Thanks for the fix!

**Comment by sdmaclea** [comment]
> It looks like the current fix is counting on us never going down the JIT write barrier path while the GC is running. Certainly I'd be very surprised if we called the JIT helper from within the GC but it feels more satisfying if long term we didn't have to rely on that kind of assumption.

As for the JIT write barrier case, we would run the patch until we left the write barrier.  Then we would switch to `single stepping` and hit this code. However, if the GC used managed code, we would stop during GC.  It could be problematic.

> For patching .NET 5 this looks like a good fix that minimizes the change in behavior from what we have right now. I'm hoping in the longer term fix for .NET 6 we could be a little more aggressive and avoid ever creating/queueing the DebuggerDataBreakpoint controller during the GC. 

As I wrote the alternative design description, I wondered whether it might actually be simpler.

I plan to amend this change to 
- fix deleting pinned handles. 
- Fix ICorDebugModule4::IsMappedLayout in debug/checked

I am going to draft the more direct approach too.  I think it might be simple enough to get through servicing.

File: src/coreclr/debug/ee/controller.cpp [hunk 4]
```diff
@@ -9054,23 +9071,6 @@ bool DebuggerDataBreakpoint::TriggerSingleStep(Thread *thread, const BYTE *ip)
...
-        // would return DPOSS_USED_WITH_NO_EVENT effectively treating the data breakpoint
-        // as a debugger exception which is intended to be ignored
-        this->DisableSingleStep();
-
-        // We are done with this controller
-        Delete();
-
-        // Don't queue this object to send an event
-        return false;
-    }
```
Comment (sdmaclea) [inline src/coreclr/debug/ee/controller.cpp]: I reverted this change to ignore a DebuggerDataBreakpoint controller when we were in GC.  The new change ignores the Data Breakpoint if the data breakpoint is hit during GC occurs.

If I left this code, the Data Breakpoint would also be ignored if we entered GC while stepping to find a debug safe point.

So the one risk is somehow we hit a data breakpoint and then enter GC suspend w/o first handling the GC breakpoint.

File: src/coreclr/debug/di/divalue.cpp [hunk 5]
```diff
@@ -4420,14 +4420,7 @@ HRESULT CordbHandleValue::Dispose()
...
     event.DisposeHandle.vmObjectHandle = vmObjHandle;
-    if (m_handleType == HANDLE_STRONG)
-    {
-        event.DisposeHandle.fStrong = TRUE;
-    }
-    else
-    {
-        event.DisposeHandle.fStrong = FALSE;
-    }
+    event.DisposeHandle.handleType = m_handleType;
```
Comment (sdmaclea) [inline src/coreclr/debug/di/divalue.cpp:4423]: @hoyosjs FYI, I should have had more curiosity when you asked about deleting pinned handles.  It turns out the implementation layers needed fixes (even though ICorDebug didn't).

File: src/coreclr/debug/di/module.cpp [hunk 6]
```diff
@@ -2758,18 +2758,20 @@ HRESULT CordbModule::GetJITCompilerFlags(DWORD *pdwFlags )
...
 
     ATT_REQUIRE_STOPPED_MAY_FAIL(pProcess);
-    PUBLIC_API_BEGIN(pProcess);
+
+    EX_TRY
     {
         hr = pProcess->GetDAC()->IsModuleMapped(m_vmModule, isMapped);
     }
-    PUBLIC_API_END(hr);
+    EX_CATCH_HRESULT(hr);
```
Comment (sdmaclea) [inline src/coreclr/debug/di/module.cpp:2775]: @hoyosjs This is @chuckries proposed fix to `CordbModule::IsMappedLayout`

**Comment by sdmaclea** [comment]
@noahfalk @cshung This changed quite a bit since the last review.  Can you take a look again.  I plan to take this to servicing.

@chuckries Initial testing for this is complete. Let me know if there is more to do.  I would like to be sure this is the last change required before merging and backporting to .NET 5.

**Comment by sdmaclea** [comment]
Code complete for 5.0.3 is Friday.  To ask for this fix in 5.0.3, this .NET 6 PR needs to be reviewed and merged.

@noahfalk @cshung @hoyosjs Please (re)review

**Comment by hoyosjs** [comment]
Looks good in terms of the controller logic. As long as Chuck doesn't find any issues with the GC notification being used  this looks fine ( I don't think this was the code path - I expect that to have been the out of proc notifications).

As for the other fix, I'm not sure where the InternalApiHolder reference was coming from, so that's a bit surprising... but if that's true the fix looks fine.

**Review by hoyosjs — APPROVED** [review]

File: src/coreclr/debug/di/module.cpp (hunk 6 above)
Comment (noahfalk) [reply src/coreclr/debug/di/module.cpp:2775]: No opposition but it was unclear how this fix relates to the data breakpoints issue? I think we should either clarify how it is connected in comments/commit text or fix it in a separate non-servicing PR.

**Review by noahfalk — APPROVED** [review]
This looks good to me. Despite being a bit larger the new factoring is easier to reason about being correct and it looks usable as-is for the .NET 6 fix.

We should clarify how that IsMappedLayout change connects or separate it if it isn't essential. Thanks!

**Comment by sdmaclea** [comment]
> We should clarify how that IsMappedLayout change connects or separate it if it isn't essential. Thanks!

I'll clarify it in the servicing PR.

It is an orthogonal fix, but w/o it it is impossible to test and verify this PR as VS crashes and restarts because of a debugger assert. While strictly not required for servicing (and may not even make the servicing bar) all testing was done with this patch.  The VS team would like the PR as it makes using debug & checked .NET runtime builds possible with VS.

If you really think we shouldn't include this in the servicing PR, I can pull it.

**Pull request merged**